File: surf-cloud-core/src/main/java/dev/slne/surf/cloud/core/netty/NettyPacketInfo.java
package dev.slne.surf.cloud.core.netty;

import dev.slne.surf.cloud.core.netty.client.NettyClientSource;
import java.util.Optional;
import org.jetbrains.annotations.Nullable;

public record NettyPacketInfo(
    NettySource source,
    Optional<ProxiedNettySource> proxiedSource
) {

  public NettyPacketInfo(NettySource source) {
    this(source, Optional.empty());
  }

  public NettyPacketInfo(NettySource source, @Nullable ProxiedNettySource proxiedSource) {
    this(source, Optional.ofNullable(proxiedSource));
  }

  public NettyClientSource clientSource() {
    return (NettyClientSource) source;
  }
}

File: surf-cloud-core/src/main/kotlin/dev/slne/surf/cloud/core/netty/protocol/packet/NettyPacketInfo.kt
package dev.slne.surf.cloud.core.netty.protocol.packet

import dev.slne.surf.cloud.api.netty.source.NettyClientSource
import dev.slne.surf.cloud.api.netty.source.NettyServerSource
import dev.slne.surf.cloud.api.netty.source.NettySource
import dev.slne.surf.cloud.api.netty.source.ProxiedNettySource
import org.jetbrains.annotations.Contract

data class NettyPacketInfo(
    val connection: Any, // TODO
    val proxiedSource: ProxiedNettySource<*>?
) {
    constructor(source: NettySource<*>) : this(source, null)

    @Contract(pure = true)
    fun asClientSource() = source as NettyClientSource

    @Contract(pure = true)
    fun asServerSource() = source as NettyServerSource

}
